add notes for n18418 (#339)
bitcoin-core-review-club/website: _posts/2021-05-19-#18418.md

---
layout: pr
date: 2021-05-19
title: "Increase OUTPUT_GROUP_MAX_ENTRIES to 100"
pr: 18418
authors: [fjahr]
components: ["wallet"]
host: glozow
status: upcoming
commit: 241e141
---

## Notes

* The way our wallet constructs transactions over time can leak information
  about its contents.  The most obvious example is we can assume that all UTXOs

> to the same address are controlled by the same person. UTXOs sent to different

Reviewer: "all UTXOs to the same address" -> "all UTXOs sent to the same address"/"all UTXOs with the same address"/"all UTXOs with the same scriptPubKey"

addresses may also be linked if they are spent together (a common heuristic
used in chain analysis).  Thus, if we're not careful, observant attackers can
link addresses to estimate our wallet balance and, if any one of our addresses
is deanonymized (e.g. we send it to an exchange, merchant, or block explorer
that knows our personal information or IP address), we might accidentally
reveal how much money we have!

* The Bitcoin Core wallet implements a few best-practice privacy techniques.
  One is avoiding the reuse of addresses when creating an invoice or change
address.  Another is grouping UTXOs into
[OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)s
by scriptPubKey and spending them together.

* However, `OutputGroup`s can grow quite large - after all, anyone can send as
  many transactions as they want to an address that we've published. It might
not make sense to fund a 0.015BTC transaction by sweeping a group of 150 inputs
worth 10BTC (not to mention the extra fees for all the unnecessary inputs).
The `OUTPUT_GROUP_MAX_ENTRIES` constant limits the number of UTXOs per

> `OutputGroup`. Within `GroupOutputs()`, if a scriptPubKey exceeds this many, we

Reviewer: "if a scriptPubKey exceeds this many" -> "if we have more than `OUTPUT_GROUP_MAX_ENTRIES` UTXOs with the same scriptPubKey"

batch them into multiple `OutputGroup`s with up to `OUTPUT_GROUP_MAX_ENTRIES`
UTXOs each.

* [PR#18418](https://github.com/bitcoin/bitcoin/pull/18418) increases
  `OUTPUT_GROUP_MAX_ENTRIES` from 10 to 100. The number 100 was suggested
during a previous review club,

> [#17824](https://bitcoincore.reviews/17824.html#l-339) This behavior change

Reviewer: ```suggestion
[#17824](https://bitcoincore.reviews/17824.html#l-339). This behavior change
```

> during a previous review club,
> [#17824](https://bitcoincore.reviews/17824.html#l-339) This behavior change

Reviewer: Perhaps:

```suggestion
during [a previous review club](https://bitcoincore.reviews/17824.html#l-339). This behavior change
```

constitutes just one line (and some adjustments to the tests), but it is ripe
with opportunities to explore how coin selection works.

> * For beginners, host recommends simply adding log statements and (don't forget

Reviewer: "host recommends" is a little bit awkward. Perhaps:

Try adding some log statements, re-compiling and then re-running the tests (Hint: test/functional/combine_logs.py to see logs, and you can assert that your logs are printed by adding with node.assert_debug_log(expected_msg=[your_log_statement]) to the functional test).

  to re-compile) then re-running the tests (Hint:
`test/functional/combine_logs.py` to see logs, and you can assert that your
logs are printed by adding `with
node.assert_debug_log(expected_msg=[your_log_statement])` to the functional
test).  Some good tests to play around with are wallet\_avoidreuse.py and

> wallet\_groups.py.  [This doc](https://github.com/fjahr/debugging_bitcoin) is a
> great resource (also authored by fjahr)!  You can also tinker some of the

Reviewer: ```suggestion
wallet\_groups.py.  The PR author, fjahr, has written an excellent [guide to debugging Bitcoin Core](https://github.com/fjahr/debugging_bitcoin) with some hints on adding logging and using debuggers. You can also tinker some of the
```

constants (maybe poke around for off-by-one errors) and see if things break!

* You may find some previous review clubs helpful:

	- Review Club [#17824](/17824) discussed the `avoid_reuse` flag.
	- Review Clubs [#17331](/17331) and [#17526](/17526) discussed coin
	  selection.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
What was your review approach?

2. What do the `avoid_reuse` wallet flag and `-avoidpartialspends` wallet
options do?  Which of the following combinations are possible / make sense?

	(A) `avoid_reuse = true, avoidpartialspends = true`

	(B) `avoid_reuse = true, avoidpartialspends = false`

	(C) `avoid_reuse = false, avoidpartialspends = true`

	(D) `avoid_reuse = false, avoidpartialspends = false`

> 3. If your wallet has 101 UTXOs of 0.01 BTC each and tries to send a payment of

Reviewer: Does this need to state that the 101 UTXOs are to the same address?

Author: woops! good point
